### PR TITLE
[SYCL] [L0] Use L0 immediate commandlists if specified via env var.

### PR DESCRIPTION
**Pull request by rdeodhar**
Use L0 immediate commandlists if explicitly specified.
This is a preliminary step towards using deviceid to detect devices on which use of immediate commandlists is preferred, and switching to using immediate commandlists on those devices automatically.

Signed-off-by: Rajiv Deodhar <rajiv.deodhar@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -829,9 +790,46 @@ pi_result _pi_device::initialize(int SubSubDeviceOrdinal,
         ZE_CALL_NOCHECK(zeDeviceGetCacheProperties,
                         (ZeDevice, &Count, &Properties));
       };
+
+  // Check device id for PVC.
+  // TODO: change mechanism for detecting PVC once L0 provides an interface.
+  ImmCommandListsPreferred = (ZeDeviceProperties->deviceId & 0xff0) == 0xbd0;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:796]: Did we request a Level-Zero API to tell when immediate submission is preferable for the device?

Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:796]: Yes, but the general feeling is that immediate commandlists will be the default everywhere. For now, they are known to work best on PVC. On other GPUs they may or may not work. Once known to work correctly on all GPUs we will remove this limited use of immediate commandlists.

**Review by smaslov-intel — COMMENTED** [review]
It appears I failed to submit my comments from a prior review, please take a look

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 2]
```diff
@@ -536,6 +552,16 @@ struct _pi_device : _pi_object {
   // Therefore it can be accessed without holding a lock on this _pi_device.
   const pi_device RootDevice;
 
+  // Whether to use immediate commandlists for queues on this device.
+  // For some devices (e.g. PVC) immediate commandlists are preferred.
+  bool ImmCommandListsPreferred;
+
+  // Return the Events scope to be used in for this device.
+  enum EventsScope EventsScope();
```
[the remark this answers is not in the document]
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.hpp]: Ok, done.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -829,9 +790,46 @@ pi_result _pi_device::initialize(int SubSubDeviceOrdinal,
...
   return PI_SUCCESS;
 }
 
+// Controls whether device-scope events are used, and how.
+enum EventsScope _pi_device::EventsScope() {
+  static const auto DeviceEventsStr =
+      std::getenv("SYCL_PI_LEVEL_ZERO_DEVICE_SCOPE_EVENTS");
+
+  // Set default based on preferred type of commandlists on this device.
+  auto Default = ImmCommandListsPreferred ? OnDemandHostVisibleProxy
```
[the remark this answers is not in the document]
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:807]: Early testing with AllHostVisible and OnDemandHostVisibleProxy was successful.
Lately, all the benchmark testing on PVC has been with OnDemandHostVisibleProxy.
I could try running the benchmarks again with AllHostVisible.

[the remark this answers is not in the document]
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:807]: I am fine with OnDemandHostVisibleProxy if it works well (also the more efficient in theory) as long as we have the user-visible env var to scale back to AllHostVisible

[the remark this answers is not in the document]
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:807]: Yes, the env var setting is still available and will override the default that was chosen based on type of commandlist.

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 4]
```diff
@@ -536,6 +552,16 @@ struct _pi_device : _pi_object {
   // Therefore it can be accessed without holding a lock on this _pi_device.
   const pi_device RootDevice;
 
+  // Whether to use immediate commandlists for queues on this device.
+  // For some devices (e.g. PVC) immediate commandlists are preferred.
+  bool ImmCommandListsPreferred;
+
+  // Return the Events scope to be used in for this device.
```
[the remark this answers is not in the document]
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.hpp:559]: With the OnDemandHostVisibleProxy, the proxy is only created when SYCL RT is waiting for an event on the host. Until these changes every device would use the same strategy, so there wouldn't be a conflict (in strategies). Could you try to create a test case that different devices (now using the different strategies) would still work together smoothly?

[the remark this answers is not in the document]
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.hpp:559]: A test for cross-device events has been created and is being used to verify the situation where one device might use standard commandlists and another immediate commandlists.

**Review by smaslov-intel — APPROVED** [review]

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.hpp:559]: I assume the test didn't reveal any issues currently, right?

[the remark this answers is not in the document]
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.hpp:559]: No issues found by the test - cross-device events work as expected.

**Pull request merged**